### ISSUE TITLE
Linear constraints

### ISSUE THREAD

**Issue by nc6** [tweag/ghc#15]
Sometimes it would be nice to ensure that a given constraint is used 0 or 1 times in a given scope.

As a motivating example, consider a DSL where we use constraints to govern when particular constructors can be inserted. Then we might have something like the following:

```
  -- | Set the label of something labellable
  Label :: Plot (Texted ': r) -> Plot r
  Text :: Has Texted r => T.Text -> Plot r
  -- | Set the title of a canvas
  Title :: Plot (Texted ': r) -> Plot r
```

We would like to ensure that the text of a title or label is set only once. The natural way to support this would seem to be to have a lollipop constraint arrow `=o`, which is treated as a linear arrow on the constraint dictionary.

**Comment by aspiwack** [comment]
That doesn't sound easy. For the moment I've assumed that constraint arrows ought to be understood as unrestricted.

Type-classe constraint typically can't be linear because they are used via record projections which are not linear. They could be affine (used 0 or 1 time) as you suggest, but I can't say that I'm looking forward to implement it.

I'll try to think about it. But it won't be a priority for quite a while.

**Comment by monoidal** [comment]
Closing in favour of https://github.com/tweag/linear-constraints/.
